presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -304,128 +290,146 @@ public void addInput(Page page)
+        {
+            if (resetPagesIndex) {
+                pagesIndex.clear();
+                updateMemoryUsage();
+                resetPagesIndex = false;
+            }
 
-    private Page extractOutput()
-    {
-        // INVARIANT: pagesIndex contains the full grouped & sorted data for one or more partitions
+            if (operatorFinishing && pendingInput == null && pagesIndex.getPositionCount() == 0) {
+                memoryContext.close();
+                return finished();
+            }
 
-        // Iterate through the positions sequentially until we have one full page
-        while (!pageBuilder.isFull()) {
-            if (partition == null || !partition.hasNext()) {
-                int partitionStart = partition == null ? 0 : partition.getPartitionEnd();
+            if (pendingInput != null) {
+                pendingInput = updatePagesIndex(pendingInput);
+                updateMemoryUsage();
+            }
 
-                if (partitionStart >= pagesIndex.getPositionCount()) {
-                    // Finished all of the partitions in the current pagesIndex
-                    partition = null;
-                    pagesIndex.clear();
+            // If we have unused input or are finishing, then we have buffered a full group
+            if (pendingInput != null || operatorFinishing) {
+                finishPagesIndex();
+                resetPagesIndex = true;
+                return ofResult(pagesIndex);
+            }
 
-                    // Try to extract more partitions from the pendingInput
-                    if (pendingInput != null && processPendingInput()) {
-                        partitionStart = 0;
-                    }
-                    else if (state == State.FINISHING) {
-                        state = State.FINISHED;
-                        // Output the remaining page if we have anything buffered
-                        if (!pageBuilder.isEmpty()) {
-                            Page page = pageBuilder.build();
-                            pageBuilder.reset();
-                            return page;
-                        }
-                        return null;
-                    }
-                    else {
-                        state = State.NEEDS_INPUT;
-                        return null;
-                    }
-                }
+            // pendingInput == null && !operatorFinishing
+            return yield();
+        }
 
-                int partitionEnd = findGroupEnd(pagesIndex, unGroupedPartitionHashStrategy, partitionStart);
-                partition = new WindowPartition(pagesIndex, partitionStart, partitionEnd, outputChannels, windowFunctions, peerGroupHashStrategy);
-                windowInfo.addPartition(partition);
+        private void updateMemoryUsage()
+        {
+            long bytes = pagesIndex.getEstimatedSize().toBytes();
+
+            if (pendingInput != null) {
+                bytes += pendingInput.getRetainedSizeInBytes();
```

Comment:
you don't have to. You are free to keep one page as undeclared allocation

Reply:
That's how it was implemented. Also, we count things like `PageBuilder` memory etc, so it makes sense to count pending page size too.